pkg/client/cache/eventqueue.go

```diff
@@ -0,0 +1,283 @@
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+)
+
+// EventQueue is a Store implementation that provides a sequence of compressed events to a consumer
+// along with event types.  This differs from the FIFO implementation in that FIFO does not provide
+// events when an object is deleted and does not provide the type of event.  Events are compressed
+// in a manner similar to FIFO, but accounting for event types and deletions.  The exact
+// compression semantics are as follows:
+//
+// 1.  If a watch.Added is enqueued with state X and a watch.Modified with state Y is received,
+//     these are compressed into (Added, Y)
+//
+// 2.  If a watch.Added is enqueued with state X and a watch.Deleted is received, these are
+//     compressed and the item is removed from the queue
+//
+// 3.  If a watch.Modified is enqueued with state X and a watch.Modified with state Y is received,
+//     these two events are compressed into (Modified, Y)
+//
+// 4.  If a watch.Modified is enqueued with state X and a watch.Deleted is received, these are
+//     compressed into (Deleted, X)
+//
+// It should be noted that the scenario where an object is deleted and re-added is not handled by
+// this type nor is it in scope; the reflector uses UIDs for the IDs passed to stores, so you will
+// never see a delete and a re-add for the same ID.
+//
+// This type maintains a backing store in order to provide the deleted state on watch.Deleted
```

Comment:
If the connection is lost, and the reflector has to re-watch, then the store gets a Replace.  

In the interim, an object may have been deleted, and you didn't see it so you'll leak memory in `store`.